onnxruntime/core/providers/opencl/opencl_utils.h

```diff
@@ -0,0 +1,412 @@
+  }
+
+#if USE_CL_CHECKED_CAST
+#define CL_CHECK_MEM_OBJECT_IS_BUFFER(mem)                                                                          \
+  {                                                                                                                 \
+    cl_mem_object_type ret;                                                                                         \
+    ORT_THROW_IF_CL_ERROR(clGetMemObjectInfo((cl_mem)mem, CL_MEM_TYPE, sizeof(cl_mem_object_type), &ret, nullptr)); \
+    ORT_ENFORCE(ret == CL_MEM_OBJECT_BUFFER, mem, " is not Buffer");                                                \
+  }
+#define CL_CHECK_MEM_OBJECT_IS_IMAGE_2D(mem)                                                                        \
+  {                                                                                                                 \
+    cl_mem_object_type ret;                                                                                         \
+    ORT_THROW_IF_CL_ERROR(clGetMemObjectInfo((cl_mem)mem, CL_MEM_TYPE, sizeof(cl_mem_object_type), &ret, nullptr)); \
+    ORT_ENFORCE(ret == CL_MEM_OBJECT_IMAGE2D, mem, " is not Image2D");                                              \
+  }
+
+#define CL_BUFFER_FROM_TENSOR(TENSOR) [&]() {                                                             \
+  cl_mem ptr = const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw())); \
+  if (ptr) {                                                                                              \
+    CL_CHECK_MEM_OBJECT_IS_BUFFER(ptr);                                                                   \
+  }                                                                                                       \
+  return ptr;                                                                                             \
+}()
+
+#define CL_IMAGE2D_FROM_TENSOR(TENSOR) [&]() {                                                            \
+  cl_mem ptr = const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw())); \
+  if (ptr) {                                                                                              \
+    CL_CHECK_MEM_OBJECT_IS_IMAGE_2D(ptr);                                                                 \
+  }                                                                                                       \
+  return ptr;                                                                                             \
+}()
+#else
+#define CL_CHECK_MEM_OBJECT_IS_BUFFER(mem)
+#define CL_CHECK_MEM_OBJECT_IS_IMAGE_2D(mem)
+#define CL_BUFFER_FROM_TENSOR(TENSOR) const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw()))
+#define CL_IMAGE2D_FROM_TENSOR(TENSOR) const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw()))
```

Comment:
const_cast should be avoided in general. if we're getting the data buffer from a Tensor shouldn't it be mutable and MutableDataRaw able to be called?

not quite following the call to remove_pointer_t followed by '*'. how is that different to using `cl_mem` directly?

Reply:
Input `Tensor` are all `const Tensor*`, then only the `DataRaw()` is available. `DataRaw()` returns a `const void*`, which is a low-level const, a pointer of arbitrary const object. `const cl_mem` is equivalent to `_cl_mem *const` which is a high-level const, a const pointer to arbitrary mutable object. So we need some tricks to use C++ casts here.

I think using C style cast is beter here XDD.

Reply:
- route 1
  1. `(const Tensor*)->DataRaw()` => `const void*`
  2. `const void*` => `const _cl_mem*` via a `static_cast`.
  3. ``const _cl_mem*`` => `_cl_mem*`, aka `cl_mem` via `const_cast`
- route 2
  1. `const Tensor*` => `Tensor*` via `const_cast`
  2. `(Tensor*)->MutableDataRaw()` => `void*`
  3. `void*` => `_cl_mem*`, aka `cl_mem` via `static_cast`

I just happens to chose the route 1 here.

Reply:
Inputs are immutable as they may be used by other nodes. Removing that const-ness seems dangerous given that. 

Why does the _cl_mem* need to be mutable? i.e. which OpenCL API requires a mutable pointer for what should be const input data?

Side note: Avoid C-style casts in C++ code. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast

Reply:
> Why does the _cl_mem* need to be mutable?
1. `_cl_mem*` is equivalent to `cl_mem`
2. All OpenCL API accept `cl_mem` without const qualifier.

So regardless of input or output. `cl_mem` is always mutable, we can only ensure it by ourself (not exactly, but I don't want to expand it here).

> Inputs are immutable as they may be used by other nodes.
In OpenCL EP, there are only one command queue, so at most one kernel is running at any given time.

Reply:
Or more clearly, at OpenCL API boundary:

| type | status | reason
|-------------|----|---
| `_cl_mem*` | OK | OpenCL API natively accept it.
| `_cl_mem* const` | OK | Top-level const, ptr is copied, const has no effect here.
| `const _cl_mem*` | Not OK | Low-level const, caused by ORT `context->Input(x)`. OpenCL API dose not accept it.
| `const _cl_mem* const` | Not OK | The low-level const here is not accepted by OpenCL API

Reply:
> In OpenCL EP, there are only one command queue, so at most one kernel is running at any given time.

It's not about whether multiple are running at once, it's about whether something changes the data in the input via the non-const pointer. That would break other nodes that use that input. 

Are there guarantees that OpenCL won't do in-place updates of the data passed in as input? If the API takes a mutable pointer it would suggest it doesn't.

Reply:
It is not OpenCL can guarantees, it is OpenCL programmers' repsonsibility. For example:
```c
__kernel void foo(
    __read_only image2d_t data,
    __write_only image2d_t output,
    __private const float min,
    __private const float max) {
  // ...
}
```

If programmer **correctly annotate** the cl_mem passed in with `__read_only` or `__write_only`, then the compiler will ensure `__read_only` image2d is not changed accidentally.